Make IPC features extendable on third-party devices

mengph opened the pull request:
Fixes #133210
Currently, tensor inter-process communication only supports cuda device. I hope to refactor this part of the code to be compatible with third-party device IPC features.

This PR makes IPC features extendable on third-party devices, and by the way fixed the bug that there is no cudaEventDestroy after calling cudaIpcOpenEventHandle.

mengph committed:
Make IPC features extendable on third-party devices #suppress-bc-linter

mengph commented:
@albanD Please re-trigger the CI test

mengph commented:
@ezyang @albanD BC-lint error because remove  rebuild_cuda_tensor, please adding label suppress-bc-linter and re-trigger the CI test.

mengph commented on torch/storage.py:
Change IPC related interfaces from only cuda to device optional

_release_ipc_counter_cuda  -> _release_ipc_counter_device
_share_cuda_ -> _share_device_
_new_shared_cuda -> _new_shared_device

mengph commented on torch/multiprocessing/reductions.py:
Change the IPC interface call, so that the tensors of third-party devices can also be serialized through the IPC feature.
rebuild_cuda_tensor -> rebuild_device_tensor

mengph commented on torch/csrc/StorageSharing.cpp:
Change the function calls implemented by IPC related interfaces, abstract the cuda runtime related calls into CUDAHooks, and call them through at::globalContext().getAcceleratorHooksInterface()

mengph commented on torch/__init__.py:
Move the implementation of get_device_module before from torch import multiprocessing so that the function call of torch.get_device_moduled can be made in reductions.py

mengph commented on test/test_torch.py:
Changes to the interface names called by the test case

mengph commented on c10/core/StorageImpl.h:
received_cuda_  -> received_device_

mengph commented on aten/src/ATen/detail/AcceleratorHooksInterface.h:
Four interfaces have been added. The implementation of these interfaces is related to the device runtime. Third-party devices can support IPC features by implementing these interfaces.

mengph commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
IPC-related cuda interfaces, here is the abstraction of the interface that calls the cuda runtime in the original StorageSharing.cpp, so that all types of devices can achieve compatibility with ipc features through interfaces such as at::globalContext().getAcceleratorHooksInterface(). :StorageShareDevice.

mengph commented:
> There's a lot of code motion in this PR which makes it annoying to review

@ezyang Sorry for the confusion. A lot of code has been changed here, but all the changes are inseparable.I added comments to each file, I hope it will be helpful for you to review the code. Looking forward to your review comments

mengph commented:
@ezyang 

> There's a lot of code motion in this PR which makes it annoying to review

Sorry for the confusion. A lot of code has been changed here, but all the changes are inseparable and have only one motion, which is to separate the device-related code in IPC through the HOOKS mechanism. I have added comments to each file, and I hope it will be helpful for everyone to review the code. I look forward to your review comments.

mengph commented:
@albanD @syed-ahmed @mikaylagawarecki @eqy @ezyang 
Is the review of this PR still in progress? Please give me some suggestions, should I reorganize and submit it?

ezyang commented:
I just need to sit down and review it. Maybe while I'm flying to PTC next week

ezyang commented on torch/multiprocessing/reductions.py:
```diff
@@ -595,9 +597,10 @@ def reduce_typed_storage_child(storage):
+    if current_device != "cpu" and storage.device.type == current_device:
         raise RuntimeError(
-            "Cannot pickle CUDA storage; try pickling a CUDA tensor instead"
+            "Cannot pickle Device storage; try pickling a Device tensor instead"
```
Better here would be to move this after the meta, and then just have this fail for everything that is not CPU. Looking at the code below, it doesn't look like it would work for non-CUDA non-CPU devices either

mengph replied:
@ezyang Modified

ezyang commented on torch/csrc/StorageSharing.cpp:
```diff
@@ -312,36 +310,36 @@ static PyObject* THPStorage_shareCuda(PyObject* self, PyObject* noargs) {
+    auto new_ref_counter = std::make_unique<char[]>(sizeof(std::string));
+    uint64_t new_ref_counter_offset;
+    bool new_event_sync_required;
+    at::globalContext().getAcceleratorHooksInterface().StorageShareDevice(
```
This one-two combo of getIpcHandleSize and then StorageShareDevice seems so incredibly overfit for CUDA's specific APIs for IPC handles; are you sure this really want to share the allocation logic with your external backend?

ezyang replied:
Yes, so concretely, I want to ask why there isn't a single function that does this all in one go and returns a tuple with all of the allocated handles as return values

mengph replied:
@ezyang This is a good suggestion. I have verified it in the latest submission. It is more reasonable to use tuple as the return value.

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
 
+void CUDAHooks::getIpcHandleSize(size_t& ipc_memory_handle_size,
+                                 size_t& ipc_event_handle_size) const{
+  // ipc_memory_handle consists of handlel of type cudaIpcMemHandle_t and
```
handle

mengph replied:
@ezyang  CUDAHooks::getIpcHandleSize is deprecated in the latest commit

ezyang commented on torch/csrc/StorageSharing.cpp:
```diff
@@ -312,36 +310,36 @@ static PyObject* THPStorage_shareCuda(PyObject* self, PyObject* noargs) {
+    ptrdiff_t offset_bytes;
+    auto new_memory_handle = std::make_unique<char[]>(ipc_memory_handle_size);
+    auto new_event_handle = std::make_unique<char[]>(ipc_event_handle_size);
+    auto new_ref_counter = std::make_unique<char[]>(sizeof(std::string));
```
This seems like a pessimization; previously we copied straight from shareIpcHandle into the Python byte string; now you do an intermediate copy through these dynamic allocations

mengph replied:
@ezyang  Modified

mengph commented:
@pytorchbot label "topic: not user facing

mengph commented:
@pytorchbot label "topic: not user facing"

mengph commented:
@ezyang Modifications have been made according to the review comments. Please re-import this pull request and trigger CI.

ezyang commented on torch/multiprocessing/reductions.py:
```diff
@@ -595,13 +597,13 @@ def reduce_typed_storage_child(storage):
+    elif storage.device.type != "cpu":
         raise RuntimeError(
-            "Cannot pickle meta storage; try pickling a meta tensor instead"
+            "Cannot pickle Device storage; try pickling a Device tensor instead"
```
nit: interpolate storage.device.type here in the error message

ezyang replied:
and then actually I guess you can just coalesce these two checks into one lol

mengph replied:
@ezyang Modified

ezyang commented on torch/multiprocessing/reductions.py:
```diff
@@ -177,8 +177,8 @@ def rebuild_cuda_tensor(
         if storage is None:
-            torch.cuda._lazy_init()
-            storage = storage_cls._new_shared_cuda(
+            torch.get_device_module()._lazy_init()
```
Don't you want to specifically initialize `storage_device`, instead of "the device module"?

mengph replied:
@ezyang Yes, I use get_device_module to lazy_init only for a specific device. Previously it was only cuda, now it is replaced with the third-party device for inference.

ezyang commented on torch/__init__.py:
```diff
@@ -2002,6 +2002,34 @@ def _manager_path():
+        raise RuntimeError(
+            f"Device '{device_module_name}' does not have a corresponding module registered as 'torch.{device_module_name}'."
+        )
+    return device_module
```
looks unchanged, ig?

mengph replied:
@ezyang Yes, the content of the function 'get_device_module' has not changed, it has just been moved before 'from torch import multiprocessing' to ensure that 'torch.get_device_module()' can be used in reduction.py

ezyang commented on torch/csrc/StorageSharing.cpp:
```diff
@@ -312,36 +310,28 @@ static PyObject* THPStorage_shareCuda(PyObject* self, PyObject* noargs) {
-      C10_CUDA_CHECK(
-          cudaIpcGetEventHandle(&ipc_event_handle, sent_data->event_));
-    }
+    auto ipc_tuple =
```
nit: you can use structured binding to bind all the tuple elements in one go

mengph replied:
@ezyang Modified

ezyang commented on torch/csrc/StorageSharing.cpp:
```diff
@@ -456,99 +446,32 @@ static PyObject* THPStorage_newSharedCuda(PyObject* _unused, PyObject* args) {
+      ref_counter_handle,
+      ref_counter_offset,
+      storage_offset_bytes,
+      data_ptr);
```
nit: It's nice to pass mutable arguments by pointer, to make it more obvious what the in-out args are

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+    size_t ipc_event_handle_size = sizeof(cudaIpcEventHandle_t);
+    std::string memory_handle(ipc_memory_handle_size, '\0');  // Using std::string instead of vector
+    std::string event_handle(ipc_event_handle_size, '\0');
+    std::string ref_counter(sizeof(std::string), '\0');
```
I'm not sure yet why you're doing it this way but it seems strange

mengph replied:
@ezyang ref_counter must have enough space for memcpy, now: 
std::string ref_counter(sizeof(std::string), '\0');  ->  std::string ref_counter(sizeof(sent_data->handle()), '\0');

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+
+
+void CUDAHooks::StorageNewSharedDevice(const c10::DeviceIndex& device,
+                                       bool& event_sync_required,
```
this does not need mutable reference

mengph replied:
@ezyang Modified

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+
+void CUDAHooks::StorageNewSharedDevice(const c10::DeviceIndex& device,
+                                       bool& event_sync_required,
+                                       std::string& s_ipc_event_handle,
```
ditto

mengph replied:
@ezyang Modified

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+void CUDAHooks::StorageNewSharedDevice(const c10::DeviceIndex& device,
+                                       bool& event_sync_required,
+                                       std::string& s_ipc_event_handle,
+                                       std::string& s_handle,
```
ditto

mengph replied:
@ezyang Modified

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+                                       bool& event_sync_required,
+                                       std::string& s_ipc_event_handle,
+                                       std::string& s_handle,
+                                       std::string& ref_counter_handle,
```
take this by value as you move out of it

mengph replied:
@ezyang Modified

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+                                       std::string& s_ipc_event_handle,
+                                       std::string& s_handle,
+                                       std::string& ref_counter_handle,
+                                       ptrdiff_t& ref_counter_offset,
```
does not need to be mutable ref

mengph replied:
@ezyang Modified

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+                                       std::string& s_handle,
+                                       std::string& ref_counter_handle,
+                                       ptrdiff_t& ref_counter_offset,
+                                       ptrdiff_t& storage_offset_bytes,
```
does not need mutable

mengph replied:
@ezyang Modified

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,132 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+                                       std::string& ref_counter_handle,
+                                       ptrdiff_t& ref_counter_offset,
+                                       ptrdiff_t& storage_offset_bytes,
+                                       c10::DataPtr& data_ptr) const {
```
You only have one return, return this by value lol

mengph replied:
@ezyang Modified

ezyang commented:
I've reviewed nearly everything but the big blocker is (1) verify that the old code and the new code still do the same thing (2) not sure what's going on with the std::strings

mengph commented:
@ezyang I have made some changes based on the latest review comments, and it can still be compiled and tested successfully. But I still have a question, which is below.

ezyang commented:
Which one?

mengph commented:
@pytorchbot label "suppress-bc-linter"

mengph commented:
> Which one?



> Which one?
I am referring to the two virtual functions StorageNewSharedDevice and StorageShareDevice added in AcceleratorHooksInterface.h. Currently, neither of these two functions provides a return statement and the return value is not void.

mengph commented:
> Which one?

@ezyang I am referring to the two virtual functions StorageNewSharedDevice and StorageShareDevice added in AcceleratorHooksInterface.h. Currently, neither of these two functions provides a return statement and the return value is not void.

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,134 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+    size_t ipc_memory_handle_size = shandle.handle.size();
+    size_t ipc_event_handle_size = sizeof(cudaIpcEventHandle_t);
+    std::string memory_handle(ipc_memory_handle_size, '\0');
+    std::string event_handle(ipc_event_handle_size, '\0');
```
Alright, I was wondering about this, but I think I understand now. Your problem is that you need a char* and length to build the Python bytes string from. But you cannot directly return shandle, as it is a CUDA specific type. But you have another problem which is shandle dies when it exits scope here, which means you can't pass its internal char* directly. So you copy it into an owning vector so that you can keep it live.

I guess the handle size is not that long. But I still feel bad about doing an extra copy here. The other approach we could take here is to have a PIMPL we return which has a virtualized destructor which can deallocate shandle after we take care of it. You know what, this is fine.

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,134 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+    std::string event_handle(ipc_event_handle_size, '\0');
+
+    uint64_t ref_counter_offset;
+    bool event_sync_required;
```
nit: default initialize this for safety

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -459,6 +462,134 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+        static_cast<torch::CudaIPCSentData*>(storage.data_ptr().get_context());
+    sent_data->set_original_ptr(std::move(old_data_ptr));
+    std::string ref_counter(sizeof(sent_data->handle()), '\0');
+    std::memcpy(ref_counter.data(), (sent_data->handle()).c_str(), sizeof(sent_data->handle()));
```
This looks wrong. You're copying sizeof(sent_data->handle()) which is the size of the string object itself, not the underlying data.

ezyang commented:
Can you give some evidence that a third party device can actually usefully implement these interfaces? They still seem terribly CUDA specific...

ezyang commented:
OK, that's good enough

mengph commented:
> OK, that's good enough

@ezyang Do you mean to follow the current PR plan?

mengph commented:
@ezyang Do you mean to follow the current PR solution? If so, please consider merging this PR. (There is still a legacy issue, which is to avoid extra copy by designing PIMPL as mentioned above. Maybe this problem can be solved in the next submission. What do you think?)

mengph committed:
process handle with PIMPL. #suppress-bc-linter

mengph commented:
@ezyang I found an example of passing handles through PIMPL in https://github.com/pytorch/pytorch/blob/main/c10/cuda/CUDACachingAllocator.cpp#L1546.
In the latest commit, three handles are processed through PIMLP, which is consistent with the above example. Do you think this is feasible? I look forward to your review.

ezyang commented:
needs some fixes

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -463,6 +466,131 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
   c10::cuda::device_synchronize();
 }
 
+std::tuple<size_t, size_t, ptrdiff_t, uint64_t, bool, c10::IPCHandlePimpl>
```
No, you've fundamentally misunderstood what exactly the pimpl is supposed to do here. You've still hard coded that this struct has to be three std::string. A true pimpl would have an abstract interface in c10 that has a virtual destructor and says nothing about what its internal state is, and then your CUDA implementation would subclass it and store precisely the data it had before.

But TBH, I preferred the old implementation (that was passing CI) because it is less abstract and there for easier for us to fix if we needed to in the future.

mengph committed:
Revert "process handle with PIMPL. #suppress-bc-linter"

This reverts commit eadf6075a95f26a0259d73aa78cd6d7a3d1795a8.

mengph commented:
@ezyang resolved conflict . Please help me trigger the CI test.

ezyang commented:
sorry, needs more rebasin

mengph commented:
@ezyang please trigger CI again.

mengph commented:
@ezyang CI passed, please merge it

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -462,6 +465,133 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+    cudaIpcOpenEventHandle(&event, *ipc_event_handle);
+    C10_CUDA_CHECK(
+        cudaStreamWaitEvent(c10::cuda::getCurrentCUDAStream(device), event, 0));
+    cudaEventDestroy(event);
```
The destroy here is new

ezyang replied:
looks right and we were leaking the event before

ezyang commented:
@pytorchbot merge

ezyang commented:
I think there's a decent chance the mac failures are caused by this PR, please PTAL

mengph commented:
> I think there's a decent chance the mac failures are caused by this PR, please PTAL
@ezyang Sorry, I haven't found out which change caused it yet. Maybe you can trigger merge ci again to eliminate occasional interference.

ezyang commented:
@pytorchbot merge -r

mengph committed:
test error in macos

mengph commented:
> Yup still failing.

I suspect that it is caused by adding torch._C._get_accelerator in reduction.py, but there is no macos environment locally. So I added a temporary submission for verification here. Please help me trigger ci

ezyang commented:
Yes, this looks better. Please fix lint

mengph commented:
> Yes, this looks better. Please fix lint
@ezyang Lint fix completed, ci passed. Please check and merge.

ezyang commented on torch/multiprocessing/reductions.py:
```diff
@@ -389,15 +360,40 @@ def reduce_tensor(tensor):
+        ),
     )
-    return (rebuild_tensor, (type(tensor), storage, metadata))
 
```
there's something new going on here that I now have to review.....

ezyang reviewed (changes requested):
bouncing review until I look

ezyang commented:
btw when you force push I have no way of seeing what the incremental changes are and I have to rereview everything

mengph commented:
> btw when you force push I have no way of seeing what the incremental changes are and I have to rereview everything

@ezyang I'm sorry to bother you. After I finally confirmed that torch._C._get_accelerator() caused an error on macos, I force pushed the code once. You can think of it as a commit to solve the error on macos: https://github.com/pytorch/pytorch/commit/2f006d65c05fe8ebd17342656862249b2cec21e4
Thank you for your reminder. It won't happen again.

ezyang commented:
Is your change to the serialization code BC breaking? It sure looks BC breaking because you changed the builder function. Don't do that.

mengph committed:
add rebuild_cuda_tensor with deprecated for BC

mengph committed:
fix deprecated

mengph commented:
> Is your change to the serialization code BC breaking? It sure looks BC breaking because you changed the builder function. Don't do that.


@ezyang  BC was triggered when I changed rebuild_cuda_tensor to rebuild_cuda_tensor. I have re-added rebuild_cuda_tensor and added the deprecated flag.

ezyang commented:
Can't land this. This include is not permissible:

```
fbcode/caffe2/aten/src/ATen/cuda/detail/CUDAHooks.cpp:22:10: fatal error: 'torch/csrc/CudaIPCTypes.h' file not found
   22 | #include <torch/csrc/CudaIPCTypes.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

ezyang commented on torch/multiprocessing/reductions.py:
```diff
@@ -152,6 +153,11 @@ def rebuild_meta_tensor(
+    "`torch.multiprocessing.reductions.rebuild_cuda_tensor` is deprecated."
+    "Please use `torch.multiprocessing.reductions.rebuild_device_tensor` instead.",
+    category=FutureWarning,
+)
```
Actually, I take it back, this is multiprocessing reductions, these should have less stringent BC reqs. Harmless to keep it as is though...

mengph commented:
> Can't land this. This include is not permissible:
> 
> ```
> fbcode/caffe2/aten/src/ATen/cuda/detail/CUDAHooks.cpp:22:10: fatal error: 'torch/csrc/CudaIPCTypes.h' file not found
>    22 | #include <torch/csrc/CudaIPCTypes.h>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
> 1 error generated.
> ```

@ezyang Do you have any good suggestions? I see that there is also a precedent of "include /torch/*" under ATen, why can't CudaIPCTypes.h be used?

ezyang commented:
...there isn't? ATen cannot include torch.

I think unfortunately you need to bounce from ATen to Torch via a separate vcall.

mengph commented:
> ...there isn't? ATen cannot include torch.
> 
> I think unfortunately you need to bounce from ATen to Torch via a separate vcall.

@ezyang Maybe I can move CudaIPCTypes.* to ATen, what do you think?

ezyang commented:
Maybe, if you don't end up pulling the rest of the kitchen sink. Historically the IPC handling all lived in torch/csrc and that made it difficult to use from ATen. Especially because some of them need access to shmmanager which is a strictly torch/csrc concept. You should scope it out.

mengph committed:
move CudsIPCTypes.cpp to ATen

mengph commented:
@ezyang I moved CudaIPCTypes.cpp to ATen, please review

ezyang commented:
It doesn't seem like the patched worked, and it needs rebasing now

mengph commented:
> 补丁似乎没有起作用，现在需要重新定基

I will fix this issue in a week.

mengph commented:
> 项目似乎没有人力，现在需要重新定基

@ezyang I'll fix this next week.

mengph committed:
fix build error

mengph commented:
@ezyang ci passed, please help review it.

mengph commented:
@ezyang ci passed, please help review it.

mengph commented:
@ezyang @albanD  ci passed, please help review it.

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -466,6 +469,133 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
 }
 
+std::tuple<size_t, size_t, ptrdiff_t, std::string, std::string, std::string, uint64_t, bool>
+CUDAHooks::StorageShareDevice(const c10::Storage& storage) const {
```
NB: invariant is storage.data() is not null

ezyang commented on torch/csrc/StorageSharing.cpp:
```diff
@@ -397,38 +386,36 @@ static PyObject* THPStorage_releaseIPCCounter(
   Py_ssize_t handle_size = 0;
   if (PyBytes_AsStringAndSize(handle, &buffer, &handle_size) == -1) {
-    TORCH_CHECK(handle_size == CUDA_IPC_HANDLE_SIZE, "incorrect handle");
+    TORCH_CHECK(false, "incorrect handle");
```
This is wrong, need to query the IPC handle size and test it here

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.cpp:
```diff
@@ -466,6 +469,133 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
+                                               std::string s_handle,
+                                               std::string ref_counter_handle,
+                                               ptrdiff_t ref_counter_offset,
+                                               ptrdiff_t storage_offset_bytes) const {
```
invariant; device guard is already set

ezyang commented on torch/csrc/StorageSharing.cpp:
```diff
@@ -456,98 +446,31 @@ static PyObject* THPStorage_newSharedCuda(PyObject* _unused, PyObject* args) {
+      THPStorage_bytesAsHandleString(_event_handle);
+  bool event_sync_required = PyObject_IsTrue(_event_sync_required);
+  if (event_sync_required && s_ipc_event_handle.empty())
     return nullptr;
```
This doesn't look exactly semantics preserving. In the old code, even if the ipc event handle was empty, if an event sync was required, we would trigger a cuda stream wait event. In the new code, you always skip the sync.

mengph committed:
Modify based on review comments 2-24
